wordpress-mobile/gutenberg-mobile
# Produce Hermes bytecode and use it on Android

Fixes #1660 

This PR adds the Hermes command line to produce the binary bytecode version of the JS bundle for Android, and sets it as the asset to be loaded by the parent app.

### Changes in this PR:

1. Introduced the `bundle:android:bytecode` yarn script. Is uses the Hermes binary to compile the `App.js` bundle into binary bytecode that Hermes' runtime can use to offer better performance on Android. This script is ran automatically as part of the `bundle` script too.
2. Have JitPack copy the bytecode file into the Android bridge artifact instead of the App.JS bundle it was bundling before.
3. Have the glue code load up `index.android.bytecode` from the assets folder. That's the code the parent app runs to load the bundle so, with this change the parent app will use the binary bytecode file.
4. Include the binary bytecode file (and the accompanying mapping file) into source control. Note the commited artifacts are not really used by any Android app anyway, only included for "parity" with the string JS bundles being there too.

### Comparison on Pixel 2XL Android 10
Without bytecode | With bytecode
-- | --
~4secs loading | ~1sec loading
![block-editor-open-slow](https://user-images.githubusercontent.com/1032913/76514707-fc415700-6460-11ea-8502-8d5026de7907.gif) | ![block-editor-open-fast](https://user-images.githubusercontent.com/1032913/76514784-13804480-6461-11ea-82f1-ed011c076087.gif)

Also measured the same post opening on a Nexus 5X (Android 7.1.2) and loading takes ~3secs.

### Notes

There are more than a few warning emitted by Hermes while compiling the current App bundle into bytecode. Will document here but now sure yet if these are actually important or not. Will probably need to go over these one-by-one to assess:

```
bundle/android/App.js:36:134: warning: the variable "Promise" was not declared in function "isBoldTextEnabled"

bundle/android/App.js:48:3047: warning: the variable "DebuggerInternal" was not declared in function "value 15#"

bundle/android/App.js:110:57098: warning: the variable "setTimeout" was not declared in function "Pi 1#"

bundle/android/App.js:110:83621: warning: the variable "__REACT_DEVTOOLS_GLOBAL_HOOK__" was not declared in function "Xl"

bundle/android/App.js:110:22512: warning: the variable "clearTimeout" was not declared in function " 657#"

bundle/android/App.js:110:89330: warning: the variable "nativeFabricUIManager" was not declared in function " 692#"

bundle/android/App.js:127:795: warning: the variable "setImmediate" was not declared in function "h 7#"

bundle/android/App.js:137:177: warning: the variable "Worker" was not declared in function " 791#"

bundle/android/App.js:152:67: warning: the variable "fetch" was not declared in function " 855#"

bundle/android/App.js:152:81: warning: the variable "Headers" was not declared in function " 855#"

bundle/android/App.js:152:97: warning: the variable "Request" was not declared in function " 855#"

bundle/android/App.js:152:114: warning: the variable "Response" was not declared in function " 855#"

bundle/android/App.js:153:1539: warning: the variable "FileReader" was not declared in function "p 12#"

bundle/android/App.js:153:1973: warning: the variable "Blob" was not declared in function " 864#"

bundle/android/App.js:153:2034: warning: the variable "FormData" was not declared in function " 864#"

bundle/android/App.js:153:2107: warning: the variable "URLSearchParams" was not declared in function " 864#"

bundle/android/App.js:153:6805: warning: the variable "XMLHttpRequest" was not declared in function " 871#"

bundle/android/App.js:197:245: warning: the variable "MessageChannel" was not declared in function " 993#"

bundle/android/App.js:275:840: warning: the variable "requestAnimationFrame" was not declared in function "value 276#"

bundle/android/App.js:325:2651: warning: the variable "clearImmediate" was not declared in function " 1512#"

bundle/android/App.js:353:1238: warning: the variable "cancelAnimationFrame" was not declared in function "componentWillUnmount 3#"

bundle/android/App.js:886:5988: warning: the property "ldquo" was set multiple times in the object definition.

bundle/android/App.js:1477:12736: warning: the variable "alert" was not declared in function "warn"

bundle/android/App.js:1841:2350: warning: the variable "Intl" was not declared in function "x 32#"
```

To test:

1. Use the WPAndroid APK produced on the PR here https://github.com/wordpress-mobile/WordPress-Android/pull/11436 and observe the start-up time of the block editor. It should be faster than before.
2. Smoke test the editor, it should work as normal.

PR submission checklist:
